tests/test_urlslug.py

```diff
@@ -0,0 +1,152 @@
+        title = "12.4 sometext&amp;moretext"
+        expected = "12-4-sometext-moretext"
+        actual = generate_slug(title)
+
+        assert expected == actual
+
+    def test_it_can_identify_chapter_num_when_not_in_section_title(self):
+        """If the chapter number isn't present in the section title,
+        it can find it in the chapter title instead.
+        """
+        book_title = "college-physics"
+        chapter_title = "1 Introduction: The Nature of Science and Physics"
+        section_title = "problems-and-exercises"
+        expected = "1-problems-and-exercises"
+        actual = generate_slug(book_title, chapter_title=chapter_title, section_title=section_title)
+
+        assert expected == actual[1]
+
+    def test_when_chapter_number_not_present_in_section_nor_chapter_titles(self):
+        """What happens when (or if) a chapter number is not present in the chapter
+        title nor the section title.
+        """
+        book_title = "college-physics"
+        chapter_title = "Introduction: The Nature of Science and Physics"
+        section_title = "problems-and-exercises"
+        expected = "problems-and-exercises"
+        actual = generate_slug(
+            book_title, chapter_title=chapter_title, section_title=section_title)
+
+        assert expected == actual[1]
+
+    def test_slug_generator_integration_test(self):
+        """Integration-level test using the Acceptance Criteria described in:
+        https://github.com/openstax/cnx/issues/348
+        """
+        books = [
+            (
+                'College Physics',
+                '<span class="os-number">12.4</span><span class="os-divider"> </span><span class="os-text">Viscosity and Laminar Flow; Poiseuille\'s Law< /span >',
+                '<span class="os-number">12</span><span class="os-divider"> </span><span class="os-text">Fluid Dynamics and Its Biological and Medical Applications</span>',
+            ), (
+                'College Physics',
+                '<span class="os-text">Problems &amp; Exercises</span>',
+                '<span class="os-number">1</span><span class="os-divider"> </span><span class="os-text">Introduction: The Nature of Science and Physics</span>',
+            ), (
+                'Biology 2e',
+                '<span class="os-number">1.1</span><span class="os-divider"> </span><span class="os-text">The Science of Biology</span>',
+                '<span class="os-number">1</span><span class="os-divider"> </span><span class="os-text">The Study of Life</span>',
+            ), (
+                'Biology 2e',
+                '<span class="os-text">Preface</span>',
+            ),
+        ]
+
+        expectations = [
+            ('college-physics', '12-4-viscosity-and-laminar-flow-poiseuille-s-law'),
+            ('college-physics', '1-problems-exercises'),
+            ('biology-2e', '1-1-the-science-of-biology'),
+            ('biology-2e', 'preface')
+        ]
+
```

Comment:
We go from returning one type `str` in some situations to a `tuple` in other situations. How about we only return a `str`?

Reply:
The reason for this would be something along the lines for functional programming expectations. While we can do this kind of dynamic return in Python, that doesn't mean we should. Python isn't strictly typed but that's not to say we should ignore those practices in usage.

Reply:
Yeah, I agree. However, I would've thought you'd want the opposite. First option: you call the function twice – once (always) for generating the chapter title and the section title. Second option: you call the function once. If you provide only one argument, then you get a tuple with one item, but if you call with 2-3 arguments, you get a tuple with 2 items every time. So I'll go with the second option: to return a tuple always... with 1 OR 2 items.

Reply:
Done.